test/unit-test/FreeRTOS_Stream_Buffer/FreeRTOS_Stream_Buffer_utest.c

```diff
@@ -0,0 +1,1108 @@
+    uint32_t temp = 0;
+
+    /* Clear everything. */
+    memset( pxLocalBuffer, 0, sizeof( StreamBuffer_t ) - sizeof( pxLocalBuffer->ucArray ) + 1024 );
+
+    /* Fill the Data (0xAB) */
+    memset( pucData, 0xAB, sizeof( pucData ) );
+
+    pxLocalBuffer->LENGTH = 1024;
+    /* Set Head at 0 for simplicity. */
+    pxLocalBuffer->uxHead = 0;
+    /* Only 500 bytes available. */
+    pxLocalBuffer->uxTail = pxLocalBuffer->uxHead + uxBytesToBeWritten + 1;
+    pxLocalBuffer->uxMid = 0;
+    pxLocalBuffer->uxFront = 0;
+
+    FreeRTOS_min_uint32_Stub( FreeRTOS_min_stub );
+
+    uxReturn = uxStreamBufferAdd( pxLocalBuffer, uxOffset, pucData, uxByteCount );
+
+    /* Only these many bytes should be written. */
+    TEST_ASSERT_EQUAL( uxBytesToBeWritten - uxOffset, uxReturn );
+
+    /* Assert that data was indeed written. */
+    TEST_ASSERT_EQUAL_MEMORY( pucData, pxLocalBuffer->ucArray + uxOffset, uxBytesToBeWritten - uxOffset );
+
+    /* Make sure that 4 next bytes are zero. 4 is not a specific number. But it would make sure
+     * that we did not overwrite pre-existing data. */
+    TEST_ASSERT_EQUAL_MEMORY( &temp, pxLocalBuffer->ucArray + uxBytesToBeWritten, sizeof( temp ) );
+
+    /* Make sure that the head is not moved since data was written at an offset. */
+    TEST_ASSERT_EQUAL( 0, pxLocalBuffer->uxHead );
+}
+
+/*
+ * @brief Test adding to the buffer when it has less space than the data.
+ *        And it has non-zero offset which causes rollover.
+ */
+void test_uxStreamBufferAdd_BufferHasLessSpaceThanData_NonZeroOffsetCausesRollover( void )
+{
+    /* Now we need to get a buffer which can store up to 1024 bytes. */
+    StreamBuffer_t * pxLocalBuffer = malloc( sizeof( StreamBuffer_t ) - sizeof( pxLocalBuffer->ucArray ) + 1024 );
+    size_t uxOffset = 100;
+    uint8_t pucData[ 512 ];
+    size_t uxByteCount = sizeof( pucData );
+    size_t uxReturn;
+    size_t uxBytesToBeWritten = 500;
+    uint32_t temp = 0;
+
+    /* Clear everything. */
+    memset( pxLocalBuffer, 0, sizeof( StreamBuffer_t ) - sizeof( pxLocalBuffer->ucArray ) + 1024 );
+
+    /* Fill the Data (0xAB) */
+    memset( pucData, 0xAB, sizeof( pucData ) );
+
+    pxLocalBuffer->LENGTH = 1024;
+    /* Set Head at almost the end for rollover to happen. */
+    pxLocalBuffer->uxHead = 1000;
+    /* Only 500 bytes available. */
+    pxLocalBuffer->uxTail = pxLocalBuffer->uxHead + uxBytesToBeWritten + 1;
+    /* Rollover adjustment. */
+    pxLocalBuffer->uxTail -= pxLocalBuffer->LENGTH;
+
+    pxLocalBuffer->uxMid = 0;
+    pxLocalBuffer->uxFront = 0;
+
+    FreeRTOS_min_uint32_Stub( FreeRTOS_min_stub );
+
+    uxReturn = uxStreamBufferAdd( pxLocalBuffer, uxOffset, pucData, uxByteCount );
+
+    /* Only these many bytes should be written. */
+    TEST_ASSERT_EQUAL( uxBytesToBeWritten - uxOffset, uxReturn );
+
+    /* Assert that data was indeed written. */
+    TEST_ASSERT_EQUAL_MEMORY( pucData, pxLocalBuffer->ucArray + 76, uxBytesToBeWritten - uxOffset );
```

Comment:
What is 76?

Reply:
Replaced all magic numbers with variables explaining what is it.